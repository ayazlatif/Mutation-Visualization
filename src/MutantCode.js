import React from 'react';
import './App.css';
import './MutantDisplay.css';
import SyntaxHighlighter from 'react-syntax-highlighter';
import { docco } from 'react-syntax-highlighter/dist/esm/styles/hljs';

/* Functional component that displays details about an individual mutant */
class MutantCode extends React.Component {
    makeCodePanel(code, lines) {
        return (
            <SyntaxHighlighter showLineNumbers
                language='python'
                style={docco}
                wrapLines='true'
                lineProps={(lineNum) => {
                    if (lines.includes(lineNum)) {
                        return { class : "mutation" };
                    }
                }}>{code}</SyntaxHighlighter>
        );
    }

    render() {
        // TODO: Extract mutated lines for mutant and original
        return (
            <div style={{ maxWidth: '100%' }}>
                <script>hljs.initHighlightingOnLoad();</script>
                <div id="container">
                    <div class="panel" id="panel1">
                        <h3>Original Code</h3>
                        {this.makeCodePanel(this.props.mutant.unmutated_output, [1])}
                    </div>
                    <div class="panel" id="panel2">
                        <h3>Mutant Code</h3>
                        {this.makeCodePanel(this.props.mutant.mutated_output, [1])}
                    </div>
                    <div id="clear"></div>
                </div>
            </div>
        );
    }
};

export default MutantCode;