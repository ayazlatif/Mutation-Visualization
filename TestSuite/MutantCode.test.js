import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
// Components
import MutantCode from '../src/MutantCode';


describe('MutantCode Test Suite', () => {
    const props = {
        mutant: {
            mutated_lineno: 17,
            mutated_output: "\ndef triple_me(x):\n    return x * 3\n\n\ndef weird_boolean(b, x):\n    if b:\n        return x / 4 + 3\n    elif x < 10:\n        return x\n    return 4",
            productive: false,
            mutation_operator: "AOD",
            equivalent: false,
            mutated_output_lineno: 12,
            unmutated_output: "\ndef triple_me(x):\n    return x * 3\n\n\ndef weird_boolean(b, x):\n    if b:\n        return x / 4 + 3\n    elif x < 10:\n        return -x\n    return 4",
            killers: [],
            unmutated_output_lineno: 12,
            mutant_name: "mutant_AOD_UnaryOp_0",
            killed: false,
            mutated_ast_node: "UnaryOp"
        }
    }
    const wrapper = shallow(<MutantCode {...props} />);
    const mountWrapper = mount(<MutantCode {...props} />);

    it('Matches snapshot', () => {
        const tree = renderer.create(<MutantCode {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Component renders without crashing providing proper JSON', () => {
        expect(wrapper.exists()).toBe(true);
        expect(mountWrapper.exists()).toBe(true);
        expect(wrapper.find('div')).toHaveLength(5)
    });

    it('Panels for original and mutant code rendered correctly', () => {
        expect(wrapper.find('.panel')).toHaveLength(2);
        expect(wrapper.find('h3')).toHaveLength(2);
        expect(wrapper.find('SyntaxHighlighter')).toHaveLength(2);
    });

});