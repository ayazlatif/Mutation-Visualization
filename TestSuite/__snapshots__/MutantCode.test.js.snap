// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MutantCode Test Suite Matches snapshot 1`] = `
<div
  style={
    Object {
      "maxWidth": "100%",
    }
  }
>
  <script>
    hljs.initHighlightingOnLoad();
  </script>
  <div
    id="container"
  >
    <div
      className="panel"
      id="panel1"
    >
      <h3>
        Original Code
      </h3>
      <pre
        style={
          Object {
            "background": "#f8f8ff",
            "color": "#000",
            "display": "block",
            "overflowX": "auto",
            "padding": "0.5em",
          }
        }
      >
        <code
          style={
            Object {
              "float": "left",
              "paddingRight": "10px",
            }
          }
        >
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            12

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            13

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            14

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            15

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            16

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            17

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            18

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            19

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            20

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            21

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            22

          </span>
        </code>
        <code>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
              
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              def
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
               
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#458",
                  "fontWeight": "bold",
                }
              }
            >
              triple_me
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#00f",
                }
              }
            >
              (x)
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
              :
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               x * 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              3
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            

          </span>
          <span
            style={Object {}}
          >
            

          </span>
          <span
            className="mutation"
            style={Object {}}
          >
            <span
              style={Object {}}
            >
              
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              def
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
               
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#458",
                  "fontWeight": "bold",
                }
              }
            >
              weird_boolean
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#00f",
                }
              }
            >
              (b, x)
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
              :
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              if
            </span>
            <span
              style={Object {}}
            >
               b:

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                      
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               x / 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              4
            </span>
            <span
              style={Object {}}
            >
               + 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              3
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              elif
            </span>
            <span
              style={Object {}}
            >
               x &lt; 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              10
            </span>
            <span
              style={Object {}}
            >
              :

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                      
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               -x

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              4
            </span>
          </span>
        </code>
      </pre>
    </div>
    <div
      className="panel"
      id="panel2"
    >
      <h3>
        Mutant Code
      </h3>
      <pre
        style={
          Object {
            "background": "#f8f8ff",
            "color": "#000",
            "display": "block",
            "overflowX": "auto",
            "padding": "0.5em",
          }
        }
      >
        <code
          style={
            Object {
              "float": "left",
              "paddingRight": "10px",
            }
          }
        >
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            12

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            13

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            14

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            15

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            16

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            17

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            18

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            19

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            20

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            21

          </span>
          <span
            className="react-syntax-highlighter-line-number"
            style={Object {}}
          >
            22

          </span>
        </code>
        <code>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
              
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              def
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
               
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#458",
                  "fontWeight": "bold",
                }
              }
            >
              triple_me
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#00f",
                }
              }
            >
              (x)
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
              :
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               x * 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              3
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            

          </span>
          <span
            style={Object {}}
          >
            

          </span>
          <span
            className="mutation"
            style={Object {}}
          >
            <span
              style={Object {}}
            >
              
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              def
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
               
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#458",
                  "fontWeight": "bold",
                }
              }
            >
              weird_boolean
            </span>
            <span
              className="hljs-function"
              style={
                Object {
                  "color": "#00f",
                }
              }
            >
              (b, x)
            </span>
            <span
              className="hljs-function"
              style={Object {}}
            >
              :
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              if
            </span>
            <span
              style={Object {}}
            >
               b:

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                      
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               x / 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              4
            </span>
            <span
              style={Object {}}
            >
               + 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              3
            </span>
            <span
              style={Object {}}
            >
              

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              elif
            </span>
            <span
              style={Object {}}
            >
               x &lt; 
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              10
            </span>
            <span
              style={Object {}}
            >
              :

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                      
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               x

            </span>
          </span>
          <span
            style={Object {}}
          >
            <span
              style={Object {}}
            >
                  
            </span>
            <span
              style={
                Object {
                  "color": "#954121",
                }
              }
            >
              return
            </span>
            <span
              style={Object {}}
            >
               
            </span>
            <span
              style={
                Object {
                  "color": "#40a070",
                }
              }
            >
              4
            </span>
          </span>
        </code>
      </pre>
    </div>
    <div
      id="clear"
    />
  </div>
</div>
`;
